From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Marko <visualpvp1@gmail.com>
Date: Sun, 5 May 2024 21:54:12 +0200
Subject: [PATCH] Testing Patch


diff --git a/src/main/java/net/minecraft/server/Achievement.java b/src/main/java/net/minecraft/server/Achievement.java
index 09b69bb3556b5edaa853ba04e45d885a4d6f7344..028a4777a7c4e60f90af21b3de1689005bf9892d 100644
--- a/src/main/java/net/minecraft/server/Achievement.java
+++ b/src/main/java/net/minecraft/server/Achievement.java
@@ -35,6 +35,10 @@ public class Achievement extends Statistic {
         this.c = achievement;
     }
 
+    private void testPatch() {
+
+    }
+
     public Achievement a() {
         this.f = true;
         return this;
