From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Marko <visualpvp1@gmail.com>
Date: Mon, 6 May 2024 00:17:58 +0200
Subject: [PATCH] Potion Effects + Other Things


diff --git a/src/main/java/net/minecraft/server/EntityTypes.java b/src/main/java/net/minecraft/server/EntityTypes.java
index 064d0b2047498b2268f2d56233ac2beb6966ac6c..a96d6892791062de7b3ec6e800b068486f100fa0 100644
--- a/src/main/java/net/minecraft/server/EntityTypes.java
+++ b/src/main/java/net/minecraft/server/EntityTypes.java
@@ -12,15 +12,15 @@ import org.apache.logging.log4j.Logger;
 
 public class EntityTypes {
 
-    private static final Logger b = LogManager.getLogger();
-    private static final Map<String, Class<? extends Entity>> c = Maps.newHashMap();
-    private static final Map<Class<? extends Entity>, String> d = Maps.newHashMap();
-    private static final Map<Integer, Class<? extends Entity>> e = Maps.newHashMap();
-    private static final Map<Class<? extends Entity>, Integer> f = Maps.newHashMap();
-    private static final Map<String, Integer> g = Maps.newHashMap();
-    public static final Map<Integer, EntityTypes.MonsterEggInfo> eggInfo = Maps.newLinkedHashMap();
-
-    private static void a(Class<? extends Entity> oclass, String s, int i) {
+    public static Logger b = LogManager.getLogger();
+    public static Map<String, Class<? extends Entity>> c = Maps.newHashMap();
+    public static Map<Class<? extends Entity>, String> d = Maps.newHashMap();
+    public static Map<Integer, Class<? extends Entity>> e = Maps.newHashMap();
+    public static Map<Class<? extends Entity>, Integer> f = Maps.newHashMap();
+    public static Map<String, Integer> g = Maps.newHashMap();
+    public static Map<Integer, EntityTypes.MonsterEggInfo> eggInfo = Maps.newLinkedHashMap();
+
+    public static void a(Class<? extends Entity> oclass, String s, int i) {
         if (EntityTypes.c.containsKey(s)) {
             throw new IllegalArgumentException("ID is already registered: " + s);
         } else if (EntityTypes.e.containsKey(Integer.valueOf(i))) {
@@ -38,7 +38,7 @@ public class EntityTypes {
         }
     }
 
-    private static void a(Class<? extends Entity> oclass, String s, int i, int j, int k) {
+    public static void a(Class<? extends Entity> oclass, String s, int i, int j, int k) {
         a(oclass, s, i);
         EntityTypes.eggInfo.put(Integer.valueOf(i), new EntityTypes.MonsterEggInfo(i, j, k));
     }
@@ -227,11 +227,11 @@ public class EntityTypes {
 
     public static class MonsterEggInfo {
 
-        public final int a;
-        public final int b;
-        public final int c;
-        public final Statistic killEntityStatistic;
-        public final Statistic e;
+        public int a;
+        public int b;
+        public int c;
+        public Statistic killEntityStatistic;
+        public Statistic e;
 
         public MonsterEggInfo(int i, int j, int k) {
             this.a = i;
diff --git a/src/main/java/org/bukkit/craftbukkit/CraftServer.java b/src/main/java/org/bukkit/craftbukkit/CraftServer.java
index bc80d9afbc538b9926b3ff67ff80b7fdcd294aed..317fc7c28e1c9219192db2154a4b7c184fe409c2 100644
--- a/src/main/java/org/bukkit/craftbukkit/CraftServer.java
+++ b/src/main/java/org/bukkit/craftbukkit/CraftServer.java
@@ -86,8 +86,8 @@ public final class CraftServer implements Server {
     private final Logger logger = Logger.getLogger("Minecraft");
     private final ServicesManager servicesManager = new SimpleServicesManager();
     private final CraftScheduler scheduler = new CraftScheduler();
-    private final SimpleCommandMap commandMap = new SimpleCommandMap(this);
-    private final SimpleHelpMap helpMap = new SimpleHelpMap(this);
+    public SimpleCommandMap commandMap = new SimpleCommandMap(this);
+    public SimpleHelpMap helpMap = new SimpleHelpMap(this);
     private final StandardMessenger messenger = new StandardMessenger();
     private final SimplePluginManager pluginManager = new SimplePluginManager(this, commandMap); // PandaSpigot
     protected final MinecraftServer console;
